Updated SOP (#225)
CDCgov/cfa-epinow2-pipeline: SOP.md

## Introduction

### The whole workflow

```mermaid
flowchart TD

   subgraph Azure Batch pools
      subgraph cfa-epinow2-latest
           myjob("set up job") -->
           run("run epinow2") -->
           write("write outputs")
      end
   end

   subgraph Azure blob storage
     azc("
       **configs**
       az://rt-epinow2-config/
       ")
     outprod("
       **production outputs**
       az://nssp-rt-v2
       ")
     outtest("
       **non-production outputs**
       az://nssp-rt-testing
       ")
     data("
       **data**
       az://nssp-etl/gold/
       ")
   end

   subgraph Azure container registry
     acr_latest("rt-epinow2-config:latest")
   end

  init[
    **configure job**
    with cfa-config-generator
   ] -- generate -->

	 configs[/
		job_id/task_1_id.json
		job_id/task_2_id.json
		...
		job_id/task_n_id.json
	 /] -- upload --> azc

   data -- pull --> myjob
   acr_latest -- pull --> myjob
   azc -- pull --> myjob
   write -- production run --> outprod
   write -- test run --> outtest
```

### Definitions

> A **job** is a unique pipeline run, which consists of a set of configuration files that fully define the run, and a set of outputs. The `job_id` a unique tag assigned by the config generator that links the configuration files to the run and the ouputs. A job may contain many tasks (e.g. a typical production job contains tasks for each of 51 jurisdictions * 2 diseases = 102 tasks, each with their own `task_id`). Each task is reproducible and defined by its own .json config file.

Reviewer: ```suggestion
A **job** is a unique pipeline run, which consists of a set of configuration files that fully define the run, and a set of outputs. The `job_id` a unique tag assigned by the config generator that links the configuration files to the run and the outputs. A job may contain many tasks (e.g. a typical production job contains tasks for each of 51 jurisdictions * 2 diseases = 102 tasks, each with their own `task_id`). Each task is reproducible and defined by its own .json config file.
```

Downstream of this pipeline, the cfa-Rt-postprocessing repo provides additional tools to parse and evaluate the model's outputs.

This may seem complicated, but once you have your dependencies and credentials set up, the whole workflow can be run with a single call to `make run-prod` at the command line. This SOP will walk your through the required setup and the underlying steps.

### Storage and related repos

- https://github.com/CDCgov/cfa-config-generator: a microservice that generates
  configuration files, and uploads them to Azure blob storage
- https://github.com/CDCent/cfa-rt-postprocessing: a repo with tools for postprocessing of Rt estimates
- `az://rt-epinow2-config/` an Azure blob storage container that holds
   configuration files
- `az://nssp-rt-v2/` an Azure blob storage container where we store production outputs
- `az://nssp-rt-testing/` a blob storage container where we store non-production outputs
- `az://nssp-rt-post-process/` a blob storage container where we store output from the post processor.

# Getting started

### Pre-requisites
1. VAP environment & Account
	- git (`sudo apt-get install git`)

> 	- docker CLI ([follow this script](https://teams.microsoft.com/l/message/19:064fb29dee654148b647dc8b9bab0782@thread.tacv2/1728396548512?tenantId=9ce70869-60db-44fd-abe8-d2767077fc8f&groupId=4f34eba7-8fcb-4a26-9b34-a434ea777f0c&parentMessageId=1728396548512&teamName=CFA-Predict&channelName=Help&createdTime=1728396548512)). See the [appendix](#appendix) if you wish to use

Author: ```suggestion
	- docker CLI ([follow this script](https://gist.github.com/jkislin/a575780249c3580a472e25c07d3fd68f#file-docker_setup-sh)). See the [appendix](#appendix) if you wish to use
```

Yes, I didn't realize there was a gist! Was hoping for a better link.

	</details><br>

	**Knowledge check:** Where can you find the outputs of a local run?

	<details>
  	<summary>Answer</summary>
  	`ls` should reveal two new local directories: `inputs/` contains a copy of
		the model inputs, and `job_id` contains the model outputs. These files are
		copied from the container to your local file system. Feel free to clean them
		up using `rm -r inputs` and `rm -r test`.
	</details>
<br>

2. The following command will test your connection to Azure Batch resources. This command will run the pipeline for a single state (NY) and run using Azure Batch resources. To track the status of the nodes and pool, open Azure Batch Explorer. The outputs will write
to the `nssp-rt-testing` container, in a directory labeled with the job_id printed
to your command line. `make test-batch`

	**Knowledge check:** Verify that you know how to track the jobs running in the
batch explorer. Locate the outputs in Azure blob storage. Explore the
contents, and open `tasks/{taks_id}/logs.txt` which is useful for debugging.

	<details>
  	<summary>Hint</summary>
  	The outputs will be written to `az:/nssp-rt-testing/{job_id}/`
	</details>

### Rt Estimation Pipeline (Production)

> If you have succesfully setup the pre-requisites and are able to run `make config` and `make run CONGIF=test/test.json` you are ready to run the entire pipeline in production `make run-prod`.

Reviewer: ```suggestion
If you have successfully setup the pre-requisites and are able to run `make config` and `make run CONGIF=test/test.json` you are ready to run the entire pipeline in production `make run-prod`.
```

To exclude a state from the initial model run (e.g. if no data were recieved), generate appropriate configs. Task exclusions are state and disease pairs (ex. 'NY:COVID-19)' that users can specify when running the config generator so that tasks for specified state disease pairs are not generated unnecessarily. Users can specify more than one state and disease pair, so long as they are separate by a ',' (ex. 'NY:COVID-19,WA:Influenza'). To do so, a user will need to first, generate the configuration files and specify the state:disease pair to exclude (see example below) and then run the pipeline for these configurations. T

1. `TIMESTAMP=$(date -u +"%Y%m%d_%H%M%S"); JOB_ID=Rt-estimation-$(echo $TIMESTAMP)`
2. `gh workflow run -R cdcgov/cfa-config-generator run-workload.yaml -f job_id=$(echo $JOB_ID) -f task_exclusions='NY:COVID-19', -f output_container='nssp-rt-v2'`
3. `make run-batch JOB=$(echo $JOB_ID)`

Note: the `JOB=$(echo $JOB_ID)` option in step 3. instructs the model
run to use the config files generated in steps 1. and 2. Without this option,
`make run-batch` will generate new configs using the default settings, and will
run using those.

##### Data Exclusions

> The process for handling data exclusions is currently being finalized adn this section will be finalized in April 2025. For the current procedure please reach out to Patrick Corbett (pyv3@cdc.gov).

Reviewer: ```suggestion
The process for handling data exclusions is currently being finalized and this section will be finalized in April 2025. For the current procedure please reach out to Patrick Corbett (pyv3@cdc.gov).
```

## Non-production runs
The config file uniquely defines a run. To kick off a non-production run, test,
or experiment, generate an appropriate set of configs.

**DO NOT** Write tests or experimental output to the production container,
`nssp-rt-v2`. Instead, please direct non-production output to `nssp-rt-testing`.
The testing container is the default output container in the config generator.

There are three ways to generate configs:

1. Use the GUI in the config generation repo
Navigate to https://github.com/CDCgov/cfa-config-generator/actions/workflows/run-workload.yaml,
click the "Run workflow" dropdown menu, specify your parameters, and hit "Run workflow".
This will write a set of configs to `az:/rt-epinow2-config/{job_id}`. Then run
`make run-batch JOB=job_id` to run the models using the configs you just generated.

![alt text](image-1.png)

2. Use the comand line
`gh workflow run -R cdcgov/cfa-config-generator run-workload.yaml -f job_id='my_job_id' -f task_exclusions='NY:COVID-19', -f output_container='nssp-rt-testing'`

> Use -f flags to specify arugments different from the defaults

Reviewer: ```suggestion
Use -f flags to specify arguments different from the defaults
```

3. (We recommend you do NOT) run `make config`.

> This will speficy the production container, `nssp-rt-v2`, as the output destination.

Reviewer: ```suggestion
This will specify the production container, `nssp-rt-v2`, as the output destination.
```

### Test Pre-requisites are Setup
#### Test Configuration Generation
1. `make config`
Running this command runs code located in the CDCgov/cfa-config-generator repository. This command creates a suite of configuration files, following the default settings for
a production run, and saves it into `az:/rt-epinow2-config/{job_id}`.

> If successful, the job id will print to the command line:

Author: ```suggestion
If successful, something like this output will print to your command line. Make a note of the `job_id`, which you will need to identify your configs (and eventually your model outputs) in Azure blob storage.
```

> If successful, the job id will print to the command line:

Author: I was trying to emphasize something slightly different but I think this does the job?

```
make config
> gh workflow run \
>   -R cdcgov/cfa-config-generator run-workload.yaml \
>   -f disease=all \
>   -f state=all \
>   -f job_id=Rt-estimation-20250318_1717845
> Created workflow_dispatch event for `run_workload.yaml` at main
```

Verify that you can locate the configuration files in Azure blob storage.

If you receive an error that you do not have the necessary permissions to run this command please reach out Agastya Mondal (ab59@cdc.gov) for assistance.

#### Test make run command

> 1. The following command will test your setup for using the `CFAEpiNow2Pipeline` package. This command will run the pipeline for a single state and disease locally (using the computing power of your VAP account). This will take a few minutes. (N.B. the test config runs a short chain. Running the sampler locally for 10,000 steps, as in a production run, would take longer.) `make run CONFIG=test/test.json`

Author: Are you sure it's not 10,000 sampling?

> 1. The following command will test your setup for using the `CFAEpiNow2Pipeline` package. This command will run the pipeline for a single state and disease locally (using the computing power of your VAP account). This will take a few minutes. (N.B. the test config runs a short chain. Running the sampler locally for 10,000 steps, as in a production run, would take longer.) `make run CONFIG=test/test.json`

Reviewer 2: https://github.com/CDCgov/cfa-config-generator/blob/a18eb539326d8eb8e28775c4d299e9d815b02b6b/src/cfa_config_generator/utils/epinow2/constants.py#L9

> 1. The following command will test your setup for using the `CFAEpiNow2Pipeline` package. This command will run the pipeline for a single state and disease locally (using the computing power of your VAP account). This will take a few minutes. (N.B. the test config runs a short chain. Running the sampler locally for 10,000 steps, as in a production run, would take longer.) `make run CONFIG=test/test.json`

Author: ```suggestion
1. The following command will test your setup for using the `CFAEpiNow2Pipeline` package. This command will run the pipeline for a single state and disease locally (using the computing power of your VAP account). This will take a few minutes. (N.B. the test config runs a short chain, and will be faster than the default settings.) `make run CONFIG=test/test.json`
```

> 1. The following command will test your setup for using the `CFAEpiNow2Pipeline` package. This command will run the pipeline for a single state and disease locally (using the computing power of your VAP account). This will take a few minutes. (N.B. the test config runs a short chain. Running the sampler locally for 10,000 steps, as in a production run, would take longer.) `make run CONFIG=test/test.json`

Author: I don't think the number is important, but I'm realizing that there are some differences between the configs stored in blob and I'm not sure where the default for this is set...
